### PR TITLE
Add e2e test for load balancing

### LINKED ISSUE #2153: Add E2E Test for Loadbalancing
*Description*:
>Describe the issue.

[optional *Relevant Links*:]
>Any extra documentation required to understand the issue.

### LINKED ISSUE #3744: e2e: Cookie based Consistent Hashing
*Description*:
>Describe the issue.

Relates to https://github.com/envoyproxy/gateway/issues/2624

[optional *Relevant Links*:]
>Any extra documentation required to understand the issue.

### PR DESCRIPTION
**Pull request by shawnh2**
**What type of PR is this?**

**What this PR does / why we need it**:

Add e2e test for:
- [x] round robin lb
- [x] consistent hash + source ip lb
- [x] consistent hash + header lb

**Which issue(s) this PR fixes**:

Fix #2153

### CODE REVIEW

File: test/e2e/base/manifests.yaml [hunk 1]
```diff
@@ -200,7 +200,7 @@ metadata:
   labels:
     app: infra-backend-v3
 spec:
-  replicas: 2
+  replicas: 4
```
Comment (zhaohuabing) [inline test/e2e/base/manifests.yaml:203]: Do we have to increase the replicas?

File: test/e2e/tests/load_balancing.go [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+			req := http.MakeRequest(t, &expectedResponse, gwAddr, "HTTP", "http")
+
+			cookies := []string{"0.0.0.0", "1.2.3.4", "4.5.6.7", "7.8.9.10", "10.11.12.13"}
+
+			for _, cookie := range cookies {
+				// Same test cookie will always hit the same endpoint.
+				var expectPodName string
+
+				for i := 0; i < sendRequests; i++ {
+					req.Headers["Set-Cookie"] = []string{fmt.Sprintf("Lb-Test=%s", cookie)}
```
Comment (shawnh2) [inline test/e2e/tests/load_balancing.go:270]: It seems that gatewayapi conformance test suite does not support setting cookie for one request, and this cookie-setting way is ref to `http.WriteCookie` in `net/http` pkg, but unfortunately, not working as well.

File: test/e2e/base/manifests.yaml (hunk 1 above)
Comment (shawnh2) [reply test/e2e/base/manifests.yaml:203]: for testing a load-balancer, just 2 endpoints seems a little insufficient, would be nice if we can add more endpoints.

rethink about this setup, maybe we should just create a deployment with replicas=4 only for LB test case, not other cases. could save some resources this way.

File: test/e2e/testdata/load_balancing_consistent_hash_cookie.yaml [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+  targetRefs:
+    - group: gateway.networking.k8s.io
+      kind: HTTPRoute
+      name: cookie-lb-route
+  loadBalancer:
+    type: ConsistentHash
+    consistentHash:
+      type: Cookie
+      cookie:
+        name: "Lb-Test"
```
Comment (arkodg) [inline test/e2e/testdata/load_balancing_consistent_hash_cookie.yaml:16]: can we add an attribute `SameSite=Strict` in here and check in client side if it is being set ?

**Review by arkodg — APPROVED** [review]
LGTM thanks !

**Review by zhaohuabing — APPROVED** [review]
LGTM. Thanks for adding the tests!

**Comment by shawnh2** [comment]
hold for a while, these test cases seem flaky.

**Comment by shawnh2** [comment]
> hold for a while, these test cases seem flaky.

all clear, PTAL

**Review by Xunzhuo — COMMENTED** [review]
Do we have any cases for Cookie based ConsistentHash?

**Comment by shawnh2** [comment]
> Do we have any cases for Cookie based ConsistentHash?

Not yet, https://github.com/envoyproxy/gateway/pull/3868#discussion_r1681162086, will be added in another PR

**Review by Xunzhuo — APPROVED** [review]

**Pull request merged**